src/offset_of.rs

```diff
@@ -66,30 +66,30 @@
 /// }
 /// ```
 #[macro_export]
-#[cfg(memoffset_constant_expression)]
+#[cfg(memoffset_maybe_uninit)]
 macro_rules! offset_of {
-    ($parent:ty, $($field:tt)+) => (unsafe {
-        let x: &'static $parent = $crate::Transmuter::<$parent> { int: 0 }.ptr;
-        $crate::Transmuter { ptr: &x.$($field)+ }.int
+    ($parent:ty, $field:tt) => (unsafe {
+        // Create an instance of the container and calculate the offset to its
+        // field. Although we are creating references to uninitialized data this
+        // is fine since we are not dereferencing them.
+        let val = $crate::mem::MaybeUninit::<$parent>::uninitialized();
+        let &$container { $field: ref f, .. } = &*val.as_ptr();
+        #[allow(unused_unsafe)]
+        let result = unsafe { (f as *const _ as *const u8).offset_from(val.as_ptr() as *const u8) };
+        result as isize
     });
 }
 
 #[macro_export]
-#[cfg(not(memoffset_constant_expression))]
+#[cfg(not(memoffset_maybe_uninit))]
 macro_rules! offset_of {
-    ($father:ty, $($field:tt)+) => ({
-        #[allow(unused_unsafe)]
-        let root: $father = unsafe { $crate::mem::uninitialized() };
-
-        let base = &root as *const _ as usize;
-
-        // Future error: borrow of packed field requires unsafe function or block (error E0133)
+    ($parent:ty, $field:tt) => ({
+        let non_null = $crate::ptr::NonNull::<$parent>::dangling();
+        let base_ptr = unsafe { non_null.as_ref() };
```

Comment:
There should be a comment here saying "yes this is UB", explaining why it is UB, and why it does this anyway.

It is UB because the reference is dangling, i.e., it does not point to actually allocated memory.